From b36fea3666a0f0717026f33692e8fa5262b201de Mon Sep 17 00:00:00 2001
From: Tomas Mraz <tomas@openssl.foundation>
Date: Fri, 23 Oct 2026 16:28:22 +0200
Subject: [PATCH] Explicitly add relevant defines to libdefault.a

Fixes https://github.com/openssl/openssl/issues/33096

Reviewed-by: Richard Levitte <levitte@openssl.org>
Reviewed-by: Dmitry Belyavskiy <beldmit@gmail.com>
Reviewed-by: Mounir Idrassi <mounir.idrassi@idrix.fr>
Merge-date: Mon Oct 26 19:09:02 2026
Merged-from: https://github.com/openssl/openssl/pull/33359
(cherry picked from commit 646bf5a6a602b267967059fd749167d9eb1d788e)
---
 crypto/camellia/build.info                   | 1 +
 crypto/chacha/build.info                     | 1 +
 crypto/des/build.info                        | 1 +
 crypto/hmac/build.info                       | 1 +
 crypto/md5/build.info                        | 1 +
 crypto/ml_kem/build.info                     | 1 +
 crypto/modes/build.info                      | 1 +
 crypto/poly1305/build.info                   | 1 +
 crypto/ripemd/build.info                     | 1 +
 crypto/sm3/build.info                        | 1 +
 crypto/sm4/build.info                        | 1 -
 providers/implementations/ciphers/build.info | 2 +-
 ssl/record/build.info                        | 2 +-
 13 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/crypto/camellia/build.info b/crypto/camellia/build.info
index f13cd8c..7a559d8 100644
--- a/crypto/camellia/build.info
+++ b/crypto/camellia/build.info
@@ -16,6 +16,7 @@ ENDIF
 
 SOURCE[../../libcrypto]=cmll_ecb.c cmll_ofb.c cmll_cfb.c cmll_ctr.c $CMLLASM
 DEFINE[../../libcrypto]=$CMLLDEF
+DEFINE[../../providers/libdefault.a]=$CMLLDEF
 
 GENERATE[cmll-x86.S]=asm/cmll-x86.pl
 DEPEND[cmll-x86.S]=../perlasm/x86asm.pl
diff --git a/crypto/chacha/build.info b/crypto/chacha/build.info
index 41c22f7..4c93c5d 100644
--- a/crypto/chacha/build.info
+++ b/crypto/chacha/build.info
@@ -39,6 +39,7 @@ ENDIF
 
 SOURCE[../../libcrypto]=$CHACHAASM
 DEFINE[../../libcrypto]=$CHACHADEF
+DEFINE[../../providers/libdefault.a]=$CHACHADEF
 
 GENERATE[chacha-x86.S]=asm/chacha-x86.pl
 GENERATE[chacha-x86_64.s]=asm/chacha-x86_64.pl
diff --git a/crypto/des/build.info b/crypto/des/build.info
index 5a13e4f..0fe2066 100644
--- a/crypto/des/build.info
+++ b/crypto/des/build.info
@@ -28,6 +28,7 @@ IF[{- !$disabled{module} && !$disabled{shared} -}]
 ENDIF
 
 DEFINE[../../libcrypto]=$DESDEF
+DEFINE[../../providers/libdefault.a]=$DESDEF
 DEFINE[../../providers/libfips.a]=$DESDEF
 
 # When all deprecated symbols are removed, libcrypto doesn't export the
diff --git a/crypto/hmac/build.info b/crypto/hmac/build.info
index 90b43cf..b8471c6 100644
--- a/crypto/hmac/build.info
+++ b/crypto/hmac/build.info
@@ -17,6 +17,7 @@ IF[{- !$disabled{asm} -}]
 ENDIF
 
 DEFINE[../../libcrypto]=$HMACDEF
+DEFINE[../../providers/libdefault.a]=$HMACDEF
 DEFINE[../../providers/libfips.a]=$HMACDEF
 
 SOURCE[../../libcrypto]=$COMMON $HMACASM
diff --git a/crypto/md5/build.info b/crypto/md5/build.info
index 9a73db3..4ae2df9 100644
--- a/crypto/md5/build.info
+++ b/crypto/md5/build.info
@@ -33,6 +33,7 @@ ENDIF
 # Implementations are now spread across several libraries, so the defines
 # need to be applied to all affected libraries and modules.
 DEFINE[../../libcrypto]=$MD5DEF
+DEFINE[../../providers/libdefault.a]=$MD5DEF
 DEFINE[../../providers/liblegacy.a]=$MD5DEF
 
 GENERATE[md5-586.S]=asm/md5-586.pl
diff --git a/crypto/ml_kem/build.info b/crypto/ml_kem/build.info
index e593701..b1b5ef0 100644
--- a/crypto/ml_kem/build.info
+++ b/crypto/ml_kem/build.info
@@ -26,4 +26,5 @@ ENDIF
 # Implementations are now spread across several libraries, so the defines
 # need to be applied to all affected libraries and modules.
 DEFINE[../../libcrypto]=$MLKEMDEF
+DEFINE[../../providers/libdefault.a]=$MLKEMDEF
 DEFINE[../../providers/libfips.a]=$MLKEMDEF
diff --git a/crypto/modes/build.info b/crypto/modes/build.info
index cd4e033..6fcd199 100644
--- a/crypto/modes/build.info
+++ b/crypto/modes/build.info
@@ -68,6 +68,7 @@ SOURCE[../../providers/libfips.a]=$COMMON
 # Implementations are now spread across several libraries, so the defines
 # need to be applied to all affected libraries and modules.
 DEFINE[../../libcrypto]=$MODESDEF
+DEFINE[../../providers/libdefault.a]=$MODESDEF
 DEFINE[../../providers/libfips.a]=$MODESDEF
 
 
diff --git a/crypto/poly1305/build.info b/crypto/poly1305/build.info
index 5c35c8e..5cf114b 100644
--- a/crypto/poly1305/build.info
+++ b/crypto/poly1305/build.info
@@ -34,6 +34,7 @@ SOURCE[../../libcrypto]=poly1305.c $POLY1305ASM
 # Implementations are now spread across several libraries, so the defines
 # need to be applied to all affected libraries and modules.
 DEFINE[../../libcrypto]=$POLY1305DEF
+DEFINE[../../providers/libdefault.a]=$POLY1305DEF
 
 GENERATE[poly1305-sparcv9.S]=asm/poly1305-sparcv9.pl
 INCLUDE[poly1305-sparcv9.o]=..
diff --git a/crypto/ripemd/build.info b/crypto/ripemd/build.info
index 17acec0..938a554 100644
--- a/crypto/ripemd/build.info
+++ b/crypto/ripemd/build.info
@@ -17,6 +17,7 @@ ENDIF
 
 SOURCE[../../libcrypto]=rmd_dgst.c rmd_one.c $RMD160ASM
 DEFINE[../../libcrypto]=$RMD160DEF
+DEFINE[../../providers/libdefault.a]=$RMD160DEF
 
 # When all deprecated symbols are removed, libcrypto doesn't export the
 # RIPEMD160 functions, so we must include them directly in liblegacy.a
diff --git a/crypto/sm3/build.info b/crypto/sm3/build.info
index e58916d..5579014 100644
--- a/crypto/sm3/build.info
+++ b/crypto/sm3/build.info
@@ -21,6 +21,7 @@ IF[{- !$disabled{sm3} -}]
 
   SOURCE[../../libcrypto]=sm3.c legacy_sm3.c $SM3ASM
   DEFINE[../../libcrypto]=$SM3DEF
+  DEFINE[../../providers/libdefault.a]=$SM3DEF
 
   GENERATE[sm3-armv8.S]=asm/sm3-armv8.pl
   INCLUDE[sm3-armv8.o]=..
diff --git a/crypto/sm4/build.info b/crypto/sm4/build.info
index 2f36b6b..940a9d9 100644
--- a/crypto/sm4/build.info
+++ b/crypto/sm4/build.info
@@ -24,7 +24,6 @@ SOURCE[../../libcrypto]= $SM4ASM sm4.c
 # Implementations are now spread across several libraries, so the defines
 # need to be applied to all affected libraries and modules.
 DEFINE[../../libcrypto]=$SM4DEF
-DEFINE[../../providers/libfips.a]=$SM4DEF
 DEFINE[../../providers/libdefault.a]=$SM4DEF
 # We only need to include the SM4DEF stuff in the legacy provider when it's a
 # separate module and it's dynamically linked with libcrypto.  Otherwise, it
diff --git a/providers/implementations/ciphers/build.info b/providers/implementations/ciphers/build.info
index d3acd42..b5cb6d9 100644
--- a/providers/implementations/ciphers/build.info
+++ b/providers/implementations/ciphers/build.info
@@ -127,7 +127,7 @@ SOURCE[$AES_GOAL]=\
         cipher_aes_cbc_hmac_sha256_etm_hw.c \
         cipher_aes_cbc_hmac_sha512_etm_hw.c \
         cipher_cts.c
-DEFINE[$AES_GOAL]=$AESXTSDEF
+DEFINE[$AES_GOAL]=$AESXTSDEF $GHASHDEF
 
 # Extra code to satisfy the FIPS and non-FIPS separation.
 # When the AES-xxx-XTS moves to legacy, cipher_aes_xts_fips.c can be removed.
diff --git a/ssl/record/build.info b/ssl/record/build.info
index 9703d16..d44c842 100644
--- a/ssl/record/build.info
+++ b/ssl/record/build.info
@@ -1,6 +1,6 @@
 SUBDIRS=methods
 
-#Needed for the multiblock code in rec_layer_s3.c
+#Needed for the multiblock code in tls_multib.c
 IF[{- !$disabled{asm} -}]
   $AESDEF_x86=AES_ASM
   $AESDEF_x86_64=AES_ASM
